tasmota/xdrv_24_buzzer.ino

```diff
@@ -32,16 +32,39 @@ struct BUZZER {
   uint8_t inverted = 0;            // Buzzer inverted flag (1 = (0 = On, 1 = Off))
   uint8_t count = 0;               // Number of buzzes
   uint8_t mode = 0;                // Buzzer mode (0 = regular, 1 = infinite, 2 = follow LED)
+  uint8_t pwm_mode = 0;            // PWM mode (0 = regular, 1 = using PWM1 output)
   uint8_t set[2];
   uint8_t duration;
   uint8_t state = 0;
 } Buzzer;
 
 /*********************************************************************************************/
 
-void BuzzerOff(void)
+void BuzzerSet(uint8_t state)
 {
-  DigitalWrite(GPIO_BUZZER, 0, Buzzer.inverted);  // Buzzer Off
+  if (Buzzer.inverted) {
+    state = !state;
+  }
+
+  if (Buzzer.pwm_mode == 1) {
+    static uint8_t last_state = 0;
+    static uint8_t current_color = 0;
+    if (last_state != state) {
+      if (state) {
```

Comment:
I don't see why you are messing with the lights structure for something that is not a light at all. Doing this way you're just destroying the Red channel of the light.

Reply:
Thanks for reviewing. Let me answer this one by one.

1. Messing with light control?
No, see xdrv_24_buzzer.ino line 95:
```
  // If we use multi-channel PWM and have a PWM1 output, we can use PWM mode for buzzer output if enabled.
  if ((Settings.flag4.buzzer_pwm_mode) &&     // SetOption102 - Enable PWM mode for buzzer, uses PWM1 if enabled
      (Settings.flag3.pwm_multi_channels) &&  // SetOption68 - Enable multi-channels PWM instead of Color PWM
      PinUsed(GPIO_PWM1)) {
      Buzzer.pwm_mode = 1;
  }
  else {
    Buzzer.pwm_mode = 0;
  }
```
The PWM buzzer mode depends on Buzzer.pwm_mode which can only be enabled if multi-channels PWM with SetOption68 is also enabled. IOW, you can't use light control and PWM buzzer at the same time.

I asked in the TasmotaUsers group how to best control on/off an individual PWM channel in multi-channel PWM mode directly, and using "light control" is the most high-level and recommended way to do it. analogWrite() did not work and was pointed out to potentially break a lot of things:

https://groups.google.com/g/sonoffusers/c/GxOUfRmHLSg/m/jR-2MawzAgAJ

If you enable multi-channel PWM, the web UI offers to turn on/off the individual channels interactively, the same way I do it here programmatically, so I wouldn't expect this to break anything and be quite save, actually.

2. Why PWM?
Piezoelectric (cheap) buzzers need an oscillating input. See
https://en.wikipedia.org/wiki/Buzzer#Piezoelectric_2

3. Can't control the PWMFrequency?
Not true - one can use PwmFrequency command: https://tasmota.github.io/docs/Commands/
The duty cycle setting for PWM1 should be left the default 50% but I found that changing it gives slightly different sound, so one can adjust it to make the buzzer sound best.

Reply:
Thanks for you answer. I don't understand why you decided that Buzzer and Lights should not be used at the same time. If we are to make this feature available to anyone, we shouldn't have this constraint.

Please just use `analogWrite()` on `GPIO_BUZZER`. This would make it much easier to understand for users, not needing to change SO68 and not conflicting with PWM lights. This way also you make sure you that Gamma correction will not interfere.

When you say `analogWrite()` did not work, I assume it is because you did it on `GPIO_PWM1` which was in direct conflict of the light management code.

As mentioned, the only caveat is that the `PWMFrequency` should be the same for lights and buzzer.

Reply:
Thanks, I did not realize that one could use use analogWrite() on GPIO_BUZZER and the platform would then apply PWM to it, I thought that a PWM output could only work on a configured GPIO_PWMx pin. Of course if that works it is the much better option. I will try this and get back to you.

Reply:
This is now changed and works well. Again, thanks for the suggestion.

Sorry it took a while because I noticed that the generated beeps are sometimes shorter, sometimes longer and I had to do testing to make sure it's not only happening when using the new frequency output but also the plain/static output for the beeper. Measuring the output, it looks like this:

![image](https://user-images.githubusercontent.com/57991555/92455575-254d6500-f1c2-11ea-8510-bc53e0030149.png)

Even visually, you see that the times when frequency/sound is created are not always equally long. When turning the frequency output off and using it in the original static mode, the same thing is happening, too, though:

![image](https://user-images.githubusercontent.com/57991555/92455819-6fcee180-f1c2-11ea-9e3a-064d4a1a9e44.png)

This is with the command "Beep -1" which does the endless generation of beeps. I don't know how the timing for the beeper module is generated, and anyway it's not related directly to this change but should rather be a bug report. I'll post it to https://github.com/arendst/Tasmota/issues but if you already have some pointers where to look for the cause I can try to look into it.

Reply:
hi, I have a lot of positive and negative experience with the PWM due to all of my shutter related drivers. And also the AC-Dimmer that resync the PWM signal to follow the slightly changes in the AC-Frequency. From my experience, the PWMfrequency and the PWM implementation in the SDK is very stable and accurate (since june 2020). Also with wifi transmission in the background. I will check the buzzer, but if this uses normal PWM functionality any out-of-sync is very unlikely.

Reply:
The PWM frequency and generation itself is rock-solid, I agree. The timing issue lies solely in the buzzer implementation. As you can see in the bug report I filed, the exact same issue is present in the current firmware when using the Buzzer command, not using PWM/frequency output at all but static on/off output.

It seems more related to how often or when the buzzer module is called or what timing reference is used to call it. As far as I observed, the 'on' phase can become shorter and the 'off' phase can become longer, but never the other way around, at least with the non-inverted buzzer.

Reply:
If it is the first I agree. There is no guarantee in TASMOTA to get every50ms control over something. Sometimes the step is just missed if there is too much load. But in this case also a PWM does not help. To shift the on/off into a RTC timer is a bit overengineered but would solve the missing 50ms loop.

Reply:
With the Buzzer -1 command, the buzzer GPIO pin is turning on/off every 100 ms indefinitely, this is what I used. In your case using the current implementation, the output would look like this:

![image](https://user-images.githubusercontent.com/57991555/92743994-5f189a00-f381-11ea-9bd7-03ab0a294bf0.png)

Every high and low phase is approx. 100 ms here. This is turning an active buzzer such as yours on and off. When using a passive piezoelectric buzzer, this won't do anything, though, because the buzzer doesn't generate the sound itself but is more like a loudspeaker so you need to feed it the frequency it should use to generate a beeping sound. That's what the new option does. In this diagram:

![image](https://user-images.githubusercontent.com/57991555/92745031-48bf0e00-f382-11ea-986d-cb56859e9fa6.png)

the solid bars are again 100 ms wide but in this case, during the 'on' phase the 50% PWM frequency signal is generated to generate the buzzer sound. You can't see the PWM signal because of the scaling.

Reply:
PWM here is not meant to solve the timing issue, maybe a misunderstanding here. This pull request is only to add the support for passive buzzers.